camel/agents/chat_agent.py

```diff
@@ -485,220 +570,217 @@ def step(
-                    response.choices[0].message.content  # type: ignore[arg-type]
-                )
+    def _needs_tool_prompt(self) -> bool:
+        r"""Determine if a tool prompt should be injected."""
+        return (
+            not self.model_type.support_native_tool_calling
+            and self.model_backend.model_config_dict.get("tools", False)
+        )
 
-                response.choices[0].message.tool_calls = [
-                    ChatCompletionMessageToolCall(
-                        id=str(uuid.uuid4()),
-                        function=Function(
-                            arguments=str(parsed_content["arguments"]).replace(
-                                "'", '"'
-                            ),
-                            name=str(parsed_content["function"]),
-                        ),
-                        type="function",
-                    )
-                ]
+    def _inject_tool_prompt(self) -> None:
+        r"""Generate and add the tool prompt to memory."""
+        tool_prompt = self._generate_tool_prompt(
+            self.model_backend.model_config_dict["tools"]
+        )
+        tool_sys_msg = BaseMessage.make_assistant_message(
+            role_name="Assistant", content=tool_prompt
+        )
+        self.update_memory(tool_sys_msg, OpenAIBackendRole.SYSTEM)
+        self.tool_prompt_added = True
 
-                # Check for external tool call
-                tool_call_request = response.choices[0].message.tool_calls[0]
-                if tool_call_request.function.name in self.external_tool_names:
-                    # if model calls an external tool, directly return the
-                    # request
-                    info = self._step_get_info(
-                        output_messages,
-                        finish_reasons,
-                        usage_dict,
-                        response_id,
-                        tool_call_records,
-                        num_tokens,
-                        tool_call_request,
-                    )
-                    return ChatAgentResponse(
-                        msgs=output_messages,
-                        terminated=self.terminated,
-                        info=info,
-                    )
+    def _handle_step(
+        self,
+        response_format: Optional[Type[BaseModel]],
+        tool_call_records: List[FunctionCallingRecord],
+        multi_step: bool = False,
+    ) -> ChatAgentResponse:
+        r"""Handles a single or multi-step interaction."""
+        external_tool_request = None
 
-                # Normal function calling
-                tool_call_records.append(
-                    self._step_tool_call_and_update(response)
+        while True:
+            try:
+                openai_messages, num_tokens = self.memory.get_context()
+            except RuntimeError as e:
+                return self._step_token_exceed(
+                    e.args[1], tool_call_records, "max_tokens_exceeded"
```

Comment:
Is this the only cause for an exception to be thrown by memory? will this show max tokens exceeded if for example a network related exception is thrown from network backed storage?

Reply:
now `create_context` in `ScoreBasedContextCreator` only raise `RuntimeError` caused by token exceed, there's no storage related error could raise

Reply:
get_context() can call retrieve(), which could have some network errors if used with a vector db over the network right? Maybe a custom token exceeded error should be used to make sure the 3rd party code doesn't raise a RuntimeError which gets mistakenly reported as the tokens being exceeded. Or if you're confident they won't raise RuntimeErrors then lgtm